x/stakeibc/keeper/icacallbacks.go

```diff
@@ -119,6 +121,52 @@ func DelegateCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack
 	return nil
 }
 
+// ----------------------------------- redemption callback ----------------------------------- //
+func (k Keeper) MarshalRedemptionCallbackArgs(ctx sdk.Context, redemptionCallback types.RedemptionCallback) []byte {
+	out, err := proto.Marshal(&redemptionCallback)
+	if err != nil {
+		k.Logger(ctx).Error(fmt.Sprintf("MarshalRedemptionCallbackArgs %v", err.Error()))
+	}
+	return out
+}
+
+func (k Keeper) UnmarshalRedemptionCallbackArgs(ctx sdk.Context, redemptionCallback []byte) types.RedemptionCallback {
+	unmarshalledDelegateCallback := types.RedemptionCallback{}
+	if err := proto.Unmarshal(redemptionCallback, &unmarshalledDelegateCallback); err != nil {
+        k.Logger(ctx).Error(fmt.Sprintf("UnmarshalRedemptionCallbackArgs %v", err.Error()))
+	}
+	return unmarshalledDelegateCallback
+}
+
+func RedemptionCallback(k Keeper, ctx sdk.Context, packet channeltypes.Packet, ack []byte, args []byte) error {
+	// QUESTION: should we check invariants here? e.g. sendMsg.FromAddress == redemptionAddress, msg type == MsgSend, etc.
+	k.Logger(ctx).Info("RedemptionCallback executing", "packet", packet, "ack", ack, "args", args)
+	// deserialize the args
+	redemptionCallback := k.UnmarshalRedemptionCallbackArgs(ctx, args)
+	k.Logger(ctx).Info(fmt.Sprintf("RedemptionCallback %v", redemptionCallback))
+	userRedemptionRecord, found := k.RecordsKeeper.GetUserRedemptionRecord(ctx, redemptionCallback.GetUserRedemptionRecordId())
+	if !found {
+		return sdkerrors.Wrap(types.ErrRecordNotFound, "user redemption record not found")
+	}
+
+	// deserialize the ack
+	_, err := k.GetTxMsgData(ctx, ack)
+	if err != nil {
+		// ack failed, set UserRedemptionRecord as claimable
+		// NOTE: we probably only want to do this if we could unmarshal the ack and it failed
+		// DO NOT MERGE THIS INy
```

Comment:
thoughts on the cleanest way to structure this code? @sampocs

Reply:
it might make sense to pull the ack unmarshalling into `OnAcknowledgmentPacket`, then calling `CallRegisteredICACallback` with a parsed ack. That way, when `CallICACallback` is executed, we already know whether we need to handle the failure / success case (and if an ack can't be parsed at all, the callback is never executed)

Reply:
I like the idea of pulling it out - but I think it makes sense to have the ack-failure branch contained in the callback? Feels like it will be specific to the message/callback type and might make sense to bundle it with the sucess-branch, rather than having a large switch statement in `OnAcknowledgementPacket`

Reply:
That might've been what you were advocating for. But it sounded like you were thinking we only call the callback if it succeeded; whereas, I'm thinking we still parse the ack ahead of time, but call the callback regardless of success/failure?

Reply:
agreed. what do you think we should pass into `CallRegisteredICACallback`?

Reply:
probably just a bool? Unless there's specific fields we'd need from each of them, but I'd assume we could just store any info we needed in the callbackdata

Reply:
per offline discussion, let's pass in the success ack (if it's nil, assume the ack failed)